```
In file included from ../../../Source/WTF/wtf/RunLoop.h:46,
                 from ../../../Source/WTF/wtf/glib/RunLoopGLib.cpp:29:
../../../Source/WTF/wtf/WeakHashSet.h: In instantiation of ‘bool WTF::WeakHashSet< <template-parameter-1-1>, <template-parameter-1-2>, <anonymous> >::contains(const U&) const [with U = WTF::Ref<WTF::ActivityObserver>; T = WTF::Observer<void(WTF::RunLoop::Event, const WTF::String&)>; WeakPtrImpl = WTF::DefaultWeakPtrImpl; WTF::EnableWeakPtrThreadingAssertions assertionsPolicy = WTF::EnableWeakPtrThreadingAssertions::Yes]’:
../../../Source/WTF/wtf/glib/RunLoopGLib.cpp:188:5:   required from here
../../../Source/WTF/wtf/WeakHashSet.h:163:32: error: ‘const class WTF::Ref<WTF::ActivityObserver>’ has no member named ‘weakImplIfExists’
  163 |         if (auto* impl = value.weakImplIfExists(); impl && *impl)
      |                          ~~~~~~^~~~~~~~~~~~~~~~
```

--- a/Source/WTF/wtf/glib/ActivityObserver.h
+++ b/Source/WTF/wtf/glib/ActivityObserver.h
@@ -21,6 +21,7 @@
 
 #if USE(GLIB_EVENT_LOOP)
 
+#include <wtf/Lock.h>
 #include <wtf/RunLoop.h>
 #include <wtf/ThreadSafeWeakPtr.h>
 
@@ -57,10 +58,16 @@ public:
 
     void stop()
     {
-        ASSERT(m_callback);
+        {
+            Locker lock { m_callbackLock };
+            if (!m_callback)
+                return;
+
+            m_callback = nullptr;
+        }
+
         if (auto runLoop = m_runLoop.get())
             runLoop->unobserveActivity(*this);
-        m_callback = nullptr;
     }
 
     uint8_t order() const { return m_order; }
@@ -68,9 +75,12 @@ public:
 
     NotifyResult notify() const
     {
-        if (m_callback)
-            return m_callback();
-        return NotifyResult::Destroyed;
+        {
+            Locker lock { m_callbackLock };
+            if (!m_callback)
+                return NotifyResult::Destroyed;
+        }
+        return m_callback();
     }
 
 private:
@@ -87,7 +97,8 @@ private:
     ThreadSafeWeakPtr<RunLoop> m_runLoop;
     uint8_t m_order { 0 };
     OptionSet<RunLoop::Activity> m_activities;
-    Callback m_callback;
+    mutable Lock m_callbackLock;
+    Callback m_callback WTF_GUARDED_BY_LOCK(m_callbackLock);
 };
 
 } // namespace WTF

--- a/Source/WTF/wtf/glib/RunLoopGLib.cpp
+++ b/Source/WTF/wtf/glib/RunLoopGLib.cpp
@@ -185,9 +185,8 @@ void RunLoop::observeActivity(const Ref<ActivityObserver>& observer)
 
 void RunLoop::unobserveActivity(const Ref<ActivityObserver>& observer)
 {
-    // Don't assert that m_activityObservers contains the observer -- it might have been
-    // removed during notifyActivity() if it was a non-repeating observer.
     Locker locker { m_activityObserversLock };
+    ASSERT(m_activityObservers.contains(observer));
     m_activityObservers.removeFirst(observer);
 }
 
@@ -208,17 +207,9 @@ void RunLoop::notifyActivity(Activity activity)
 
     // Notify the activity observers, without holding a lock - as mutations
     // to the activity observers are allowed.
-    ActivityObservers observersToBeInvalidated;
     for (Ref observer : observersToBeNotified) {
         if (observer->notify() == ActivityObserver::NotifyResult::Stop)
-            observersToBeInvalidated.append(observer);
-    }
-
-    // Invalidation needs to happen _after_ dispatching all notifications.
-    if (!observersToBeInvalidated.isEmpty()) {
-        Locker locker { m_activityObserversLock };
-        for (Ref observer : observersToBeInvalidated)
-            m_activityObservers.removeFirst(observer);
+            observer->stop();
     }
 }
 

